File: UIKit and Animations/WWDC 2013/Customizing Your App's Appearance for iOS 7/Sample_Code/TicTacToeApp/TicTacToe/TTTMessagesViewController.m
/*
     File: TTTMessagesViewController.m
 Abstract: 
 
  Version: 1.0
 
 Disclaimer: IMPORTANT:  This Apple software is supplied to you by Apple
 Inc. ("Apple") in consideration of your agreement to the following
 terms, and your use, installation, modification or redistribution of
 this Apple software constitutes acceptance of these terms.  If you do
 not agree with these terms, please do not use, install, modify or
 redistribute this Apple software.
 
 In consideration of your agreement to abide by the following terms, and
 subject to these terms, Apple grants you a personal, non-exclusive
 license, under Apple's copyrights in this original Apple software (the
 "Apple Software"), to use, reproduce, modify and redistribute the Apple
 Software, with or without modifications, in source and/or binary forms;
 provided that if you redistribute the Apple Software in its entirety and
 without modifications, you must retain this notice and the following
 text and disclaimers in all such redistributions of the Apple Software.
 Neither the name, trademarks, service marks or logos of Apple Inc. may
 be used to endorse or promote products derived from the Apple Software
 without specific prior written permission from Apple.  Except as
 expressly stated in this notice, no other rights or licenses, express or
 implied, are granted by Apple herein, including but not limited to any
 patent rights that may be infringed by your derivative works or by other
 works in which the Apple Software may be incorporated.
 
 The Apple Software is provided by Apple on an "AS IS" basis.  APPLE
 MAKES NO WARRANTIES, EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION
 THE IMPLIED WARRANTIES OF NON-INFRINGEMENT, MERCHANTABILITY AND FITNESS
 FOR A PARTICULAR PURPOSE, REGARDING THE APPLE SOFTWARE OR ITS USE AND
 OPERATION ALONE OR IN COMBINATION WITH YOUR PRODUCTS.
 
 IN NO EVENT SHALL APPLE BE LIABLE FOR ANY SPECIAL, INDIRECT, INCIDENTAL
 OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 INTERRUPTION) ARISING IN ANY WAY OUT OF THE USE, REPRODUCTION,
 MODIFICATION AND/OR DISTRIBUTION OF THE APPLE SOFTWARE, HOWEVER CAUSED
 AND WHETHER UNDER THEORY OF CONTRACT, TORT (INCLUDING NEGLIGENCE),
 STRICT LIABILITY OR OTHERWISE, EVEN IF APPLE HAS BEEN ADVISED OF THE
 POSSIBILITY OF SUCH DAMAGE.
 
 Copyright (C) 2013 Apple Inc. All Rights Reserved.
 
 
 Copyright © 2013 Apple Inc. All rights reserved.
 WWDC 2013 License
 
 NOTE: This Apple Software was supplied by Apple as part of a WWDC 2013
 Session. Please refer to the applicable WWDC 2013 Session for further
 information.
 
 IMPORTANT: This Apple software is supplied to you by Apple Inc.
 ("Apple") in consideration of your agreement to the following terms, and
 your use, installation, modification or redistribution of this Apple
 software constitutes acceptance of these terms. If you do not agree with
 these terms, please do not use, install, modify or redistribute this
 Apple software.
 
 In consideration of your agreement to abide by the following terms, and
 subject to these terms, Apple grants you a non-exclusive license, under
 Apple's copyrights in this original Apple software (the "Apple
 Software"), to use, reproduce, modify and redistribute the Apple
 Software, with or without modifications, in source and/or binary forms;
 provided that if you redistribute the Apple Software in its entirety and
 without modifications, you must retain this notice and the following
 text and disclaimers in all such redistributions of the Apple Software.
 Neither the name, trademarks, service marks or logos of Apple Inc. may
 be used to endorse or promote products derived from the Apple Software
 without specific prior written permission from Apple. Except as
 expressly stated in this notice, no other rights or licenses, express or
 implied, are granted by Apple herein, including but not limited to any
 patent rights that may be infringed by your derivative works or by other
 works in which the Apple Software may be incorporated.
 
 The Apple Software is provided by Apple on an "AS IS" basis. APPLE MAKES
 NO WARRANTIES, EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION THE
 IMPLIED WARRANTIES OF NON-INFRINGEMENT, MERCHANTABILITY AND FITNESS FOR
 A PARTICULAR PURPOSE, REGARDING THE APPLE SOFTWARE OR ITS USE AND
 OPERATION ALONE OR IN COMBINATION WITH YOUR PRODUCTS.
 
 IN NO EVENT SHALL APPLE BE LIABLE FOR ANY SPECIAL, INDIRECT, INCIDENTAL
 OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 INTERRUPTION) ARISING IN ANY WAY OUT OF THE USE, REPRODUCTION,
 MODIFICATION AND/OR DISTRIBUTION OF THE APPLE SOFTWARE, HOWEVER CAUSED
 AND WHETHER UNDER THEORY OF CONTRACT, TORT (INCLUDING NEGLIGENCE),
 STRICT LIABILITY OR OTHERWISE, EVEN IF APPLE HAS BEEN ADVISED OF THE
 POSSIBILITY OF SUCH DAMAGE.
 
 EA1002
 5/3/2013
 */

#import "TTTMessagesViewController.h"

#import "TTTMessage.h"
#import "TTTMessageServer.h"
#import "TTTNewMessageViewController.h"

@interface TTTMessagesViewController ()
@property (strong, nonatomic) TTTProfile *profile;
@end

@interface TTTMessageTableViewCell : UITableViewCell
@property (strong, nonatomic) UIButton *replyButton;
@property (nonatomic) BOOL showReplyButton;
@end

@implementation TTTMessagesViewController {
    TTTMessage *_selectedMessage;
}

+ (UIViewController *)viewControllerWithProfile:(TTTProfile *)profile profileURL:(NSURL *)profileURL;
{
    TTTMessagesViewController *controller = [[self alloc] init];
    controller.profile = profile;
    UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:controller];
    return navController;
}

- (id)init
{
    self = [super initWithStyle:UITableViewStylePlain];
    if (self) {
        self.title = NSLocalizedString(@"Messages", @"Messages");
        self.tabBarItem.image = [UIImage imageNamed:@"messagesTab"];
        self.tabBarItem.selectedImage = [UIImage imageNamed:@"messagesTabSelected"];
        
        self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCompose target:self action:@selector(newMessage:)];
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didAddMessages:) name:TTTMessageServerDidAddMessagesNotification object:[TTTMessageServer sharedMessageServer]];
        
        self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:nil style:UIBarButtonItemStylePlain target:self action:@selector(favorite:)];
        [self updateFavoriteButton];
    }
    return self;
}

- (id)initWithStyle:(UITableViewStyle)style
{
    self = [self init];
    return self;
}

static NSString *CellIdentifier = @"Cell";

- (void)viewDidLoad
{
    [super viewDidLoad];
    [self.tableView registerClass:[TTTMessageTableViewCell class] forCellReuseIdentifier:CellIdentifier];
}

- (void)newMessage:(id)sender
{
    TTTNewMessageViewController *controller = [[TTTNewMessageViewController alloc] init];
    controller.profile = self.profile;
    [controller presentFromViewController:self];
}

- (void)didAddMessages:(NSNotification *)notification
{
    NSArray *addedIndexes = notification.userInfo[TTTMessageServerAddedMessageIndexesUserInfoKey];
    NSMutableArray *addedIndexPaths = [NSMutableArray array];
    for (NSNumber *indexValue in addedIndexes) {
        NSIndexPath *indexPath = [NSIndexPath indexPathForRow:[indexValue integerValue] inSection:0];
        [addedIndexPaths addObject:indexPath];
    }
    [self.tableView insertRowsAtIndexPaths:addedIndexPaths withRowAnimation:UITableViewRowAnimationAutomatic];
}

- (void)favorite:(id)sender
{
    BOOL favorite = [[TTTMessageServer sharedMessageServer] isFavoriteMessage:_selectedMessage];
    favorite = !favorite;
    [[TTTMessageServer sharedMessageServer] setFavorite:favorite forMessage:_selectedMessage];
    [self updateFavoriteButton];
}

- (void)updateFavoriteButton
{
    BOOL favorite = NO;
    if (_selectedMessage) {
        favorite = [[TTTMessageServer sharedMessageServer] isFavoriteMessage:_selectedMessage];
    }
    
    UIImage *image;
    if (favorite) {
        image = [[UIImage imageNamed:@"favorite"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    } else {
        image = [UIImage imageNamed:@"favoriteUnselected"];
    }
    self.navigationItem.leftBarButtonItem.image = image;
    self.navigationItem.leftBarButtonItem.enabled = (_selectedMessage != nil);
}

#pragma mark - Table view

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [[TTTMessageServer sharedMessageServer] numberOfMessages];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return [tableView dequeueReusableCellWithIdentifier:CellIdentifier forIndexPath:indexPath];
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(TTTMessageTableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath
{
    TTTMessage *message = [[TTTMessageServer sharedMessageServer] messageAtIndex:indexPath.row];
    cell.textLabel.text = [NSString stringWithFormat:@"%@", message.text];
    cell.imageView.image = [TTTProfile smallImageForIcon:message.icon];
    
    if (!cell.replyButton) {
        UIButton *replyButton = [UIButton buttonWithType:UIButtonTypeSystem];
        [replyButton addTarget:self action:@selector(newMessage:) forControlEvents:UIControlEventTouchUpInside];
        [replyButton setImage:[UIImage imageNamed:@"reply"] forState:UIControlStateNormal];
        [replyButton sizeToFit];
        cell.replyButton = replyButton;
    }
    
    BOOL isSelected = [tableView.indexPathForSelectedRow isEqual:indexPath];
    [cell setShowReplyButton:isSelected];
}

- (void)tableView:(UITableView *)tableView didDeselectRowAtIndexPath:(NSIndexPath *)indexPath
{
    TTTMessageTableViewCell *cell = (TTTMessageTableViewCell *)[tableView cellForRowAtIndexPath:indexPath];
    [cell setShowReplyButton:NO];
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    TTTMessage *message = [[TTTMessageServer sharedMessageServer] messageAtIndex:indexPath.row];
    TTTMessageTableViewCell *cell = (TTTMessageTableViewCell *)[tableView cellForRowAtIndexPath:indexPath];
    if (_selectedMessage == message) {
        [self.tableView deselectRowAtIndexPath:indexPath animated:YES];
        _selectedMessage = nil;
        [cell setShowReplyButton:NO];
    } else {
        _selectedMessage = message;
        [cell setShowReplyButton:YES];
    }
    [self updateFavoriteButton];
}

@end

@implementation TTTMessageTableViewCell

- (void)setShowReplyButton:(BOOL)value
{
    if (_showReplyButton != value) {
        _showReplyButton = value;
        if (_showReplyButton) {
            self.accessoryView = self.replyButton;
        } else {
            self.accessoryView = nil;
        }
    }
}

@end

File: UIKit and Animations/WWDC 2013/Customizing Your App's Appearance for iOS 7/Sample_Code/TicTacToeApp_Swift/TicTacToeApp_Swift/Controller/TTTMessagesViewController.swift
//
//  TTTMessagesViewController.swift
//  TicTacToeApp_Swift
//
//  Created by Seyed Samad Gholamzadeh on 7/12/18.
//  Copyright © 2018 Seyed Samad Gholamzadeh. All rights reserved.
//

import UIKit

class TTTMessagesViewController: UITableViewController {

	var profile: TTTProfile!
	var selectedMessage: TTTMessage!
	
	let CellIdentifier = "Cell"
	
	static func viewController(with profile: TTTProfile, profileURL: URL) -> UIViewController {
		
		let controller: TTTMessagesViewController = TTTMessagesViewController()
		controller.profile = profile
		let navController = UINavigationController(rootViewController: controller)
		return navController
	}
	
	init() {
		super.init(style: .plain)
		
		self.title = NSLocalizedString("Messages", comment: "Messages")
		self.tabBarItem.image = UIImage(named: "messagesTab")
		self.tabBarItem.selectedImage = UIImage(named: "messagesTabSelected")
		self.navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .compose, target: self, action: #selector(newMessage(_:)))
		NotificationCenter.default.addObserver(self, selector: #selector(didAddMessages(_:)), name: NSNotification.Name(rawValue: TTTMessageServerDidAddMessagesNotification), object: TTTMessageServer.shared)
		self.navigationItem.leftBarButtonItem = UIBarButtonItem(image: nil, style: .plain, target: self, action: #selector(favorite(_:)))
		self.updateFavoriteButton()
	}
	
	override convenience init(style: UITableViewStyle) {
		self.init()
	}
	
	required init?(coder aDecoder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}
	
    override func viewDidLoad() {
        super.viewDidLoad()

		self.tableView.register(TTTMessageTableViewCell.self, forCellReuseIdentifier: CellIdentifier)
	}
	
//	override func viewDidAppear(_ animated: Bool) {
//		super.viewDidAppear(animated)
//		self.view.window!.tintAdjustmentMode = .dimmed
//	}
	
	@objc func newMessage(_ sender: Any) {
		let controller = TTTNewMessageViewController()
		controller.profile = self.profile
		controller.present(from: self)
	}
	
	@objc func didAddMessages(_ notification: Notification) {
		let addedIndexes = notification.userInfo![TTTMessageServerAddedMessageIndexesUserInfoKey] as! [Int]
		var addedIndexPaths: [IndexPath] = []
		
		for indexValue in addedIndexes {
			let indexPath = IndexPath(row: indexValue, section: 0)
			addedIndexPaths.append(indexPath)
		}
		
		self.tableView.insertRows(at: addedIndexPaths, with: .automatic)
	}

	@objc func favorite(_ sender: Any) {
		var favorite = TTTMessageServer.shared.isFavorite(self.selectedMessage)
		favorite = !favorite
		TTTMessageServer.shared.setFavorite(favorite, for: self.selectedMessage)
		self.updateFavoriteButton()
	}

	func updateFavoriteButton() {
		var isFavorite = false
		
		if self.selectedMessage != nil {
			isFavorite = TTTMessageServer.shared.isFavorite(self.selectedMessage)
		}
		
		let image: UIImage
		if isFavorite {
			image = UIImage(named: "favorite")!.withRenderingMode(.alwaysOriginal)
		}
		else {
			image = UIImage(named: "favoriteUnselected")!
		}
		
		self.navigationItem.leftBarButtonItem?.image = image
		self.navigationItem.leftBarButtonItem?.isEnabled = (self.selectedMessage != nil)
	}


    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        // #warning Incomplete implementation, return the number of sections
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        // #warning Incomplete implementation, return the number of rows
        return TTTMessageServer.shared.numberOfMessages
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: CellIdentifier, for: indexPath)

        // Configure the cell...

        return cell
    }
	
	override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
		let message = TTTMessageServer.shared.message(at: indexPath.row)
		cell.textLabel?.text = message.text
		cell.imageView?.image = TTTProfile.smallImage(for: message.icon)
		
		if (cell as! TTTMessageTableViewCell).replyButton == nil {
			let replyButton = UIButton(type: .system)
			replyButton.addTarget(self, action: #selector(newMessage(_:)), for: .touchUpInside)
			replyButton.setImage(UIImage(named: "reply"), for: .normal)
			replyButton.sizeToFit()
			(cell as! TTTMessageTableViewCell).replyButton = replyButton
		}
		
		let isSelected = self.tableView.indexPathForSelectedRow == indexPath
		(cell as! TTTMessageTableViewCell).showReplyButton = isSelected
	}

	override func tableView(_ tableView: UITableView, didDeselectRowAt indexPath: IndexPath) {
			let cell = tableView.cellForRow(at: indexPath) as! TTTMessageTableViewCell
		cell.showReplyButton = false
	}
	
	override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
		let message = TTTMessageServer.shared.message(at: indexPath.row)
		let cell = tableView.cellForRow(at: indexPath) as! TTTMessageTableViewCell
		
		if self.selectedMessage == message {
			self.tableView.deselectRow(at: indexPath, animated: true)
			self.selectedMessage = nil
			cell.showReplyButton = false
		}
		else {
			self.selectedMessage = message
			cell.showReplyButton = true
		}
		self.updateFavoriteButton()
	}
	
}

class TTTMessageTableViewCell: UITableViewCell {
	var replyButton: UIButton!
	var showReplyButton: Bool = false {
		didSet {
			self.accessoryView = showReplyButton ? self.replyButton : nil
		}
	}
	
	
}
